src/server/ua_server.c

```diff
@@ -544,65 +678,174 @@ UA_Server_removeCallback(UA_Server *server, UA_UInt64 callbackId) {
-        if(binaryProtocolManager) {
-            binaryProtocolManager->notifyState = notifySecureChannelsStopped;
-            binaryProtocolManager->stop(server, binaryProtocolManager);
-        }
+    /* Add to the delayed callback list. Will be cleaned up in the next iteration. */
+    UA_DelayedCallback *dc = &transaction->dc;
+    dc->callback = sercureChannel_delayedClose;
+    dc->application = NULL;
+    dc->context = server;
+
+    UA_EventLoop *el = server->config.eventLoop;
+    el->addDelayedCallback(el, dc);
+
+    return UA_STATUSCODE_GOOD;
+
+cleanup:
+    UA_GDSTransaction_clear(transaction);
+    transaction = NULL;
+
+    return retval;
+}
+
+UA_StatusCode
+UA_Server_updateCertificate(UA_Server *server,
+                            const UA_NodeId certificateGroupId,
+                            const UA_NodeId certificateTypeId,
+                            const UA_ByteString *certificate,
+                            const UA_ByteString *issuerCertificates,
+                            const size_t issuerCertificatesSize,
+                            const UA_ByteString *privateKey,
+                            const UA_String *privateKeyFormat) {
+    UA_CHECK(server && certificate, return UA_STATUSCODE_BADINTERNALERROR);
+
+    /* The server currently only supports the DefaultApplicationGroup */
+    UA_NodeId defaultApplicationGroup = UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_CERTIFICATEGROUPS_DEFAULTAPPLICATIONGROUP);
+    if(!UA_NodeId_equal(&certificateGroupId, &defaultApplicationGroup))
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* The server currently only supports the following certificate types */
+    UA_NodeId certTypRsaMin = UA_NODEID_NUMERIC(0, UA_NS0ID_RSAMINAPPLICATIONCERTIFICATETYPE);
+    UA_NodeId certTypRsaSha256 = UA_NODEID_NUMERIC(0, UA_NS0ID_RSASHA256APPLICATIONCERTIFICATETYPE);
+    if(!UA_NodeId_equal(&certificateTypeId, &certTypRsaMin) &&
+       !UA_NodeId_equal(&certificateTypeId, &certTypRsaSha256))
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    if(!server->transaction) {
+        server->transaction = UA_GDSTransaction_new(server->adminSession.sessionId);
     }
 
-    size_t i = 0;
-    while(i < server->config.endpointsSize) {
-        UA_EndpointDescription *ed = &server->config.endpoints[i];
-        if(UA_ByteString_equal(&ed->serverCertificate, oldCertificate)) {
-            UA_String_clear(&ed->serverCertificate);
-            UA_String_copy(newCertificate, &ed->serverCertificate);
-            UA_SecurityPolicy *sp = getSecurityPolicyByUri(server,
-                            &server->config.endpoints[i].securityPolicyUri);
-            UA_CHECK_MEM(sp, return UA_STATUSCODE_BADINTERNALERROR);
-            sp->updateCertificateAndPrivateKey(sp, *newCertificate, *newPrivateKey);
-        }
-        i++;
+    UA_StatusCode retval = UA_GDSTransaction_addCertificateInfo(server->transaction, certificateGroupId,
```

Comment:
This looks odd. As far as I know, UA_Server_updateCertificate can be invoked either via C-API or via OPC UA. For the former case I somehow understand why server->adminSession is used. But for the latter, the OPC UA session ID must be used and of course be checked before adding changes to the transaction.

Will the transaction logic be added in a later PR and does this one only introduces some architectural foundation?

Reply:
When the updateCertificate function is called via OPC UA, a new transaction with the appropriate sessionId is created in the information model method (see https://github.com/open62541/open62541/pull/6515/files#diff-ea8abd26b36279bc52bb0e4dc60d56ecbfc041a9180c57e70eb78bfbcd504a02).

Reply:
Oh well, I did not see that the other PR also does some transaction stuff too. It seems, that both PRs depend on each other which I did not expect to see here. So sorry for the noise again. Still I have problems understanding why the two PRs are structured that way...